### PR TITLE
fix: Connection timeout for git connect when the git repo is behind a firewall

### LINKED ISSUE #12721: [Bug]: Timeout when git ssh url has a different domain than the https url
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

Setting up git sync fails when the SSH port of the git server is exposed on a different domain than the HTTPS port.

### Steps To Reproduce

Requirement: a git server that has an SSH port on domainA and an HTTPS port on domainB

When trying to set up git sync, enter the SSH url. Try to proceed, but the form will hang. In the backend you will find a timeout:

```
  - Process: JGIT status, Time elapsed: 729ms
  - boundedElastic-9: Switching to the branch master
  - Going to commit application 623d9c7ae50aa1038d108e57, branch : master
  - Error while checking if the repo is private:
 java.net.ConnectException: Connection timed out (Connection timed out)
     at java.base/java.net.PlainSocketImpl.socketConnect(Native Method)
     at java.base/java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:412)
     at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:255)
     at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:237)
     at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
     at java.base/java.net.Socket.connect(Socket.java:609)
     at java.base/sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:300)
     at java.base/sun.security.ssl.BaseSSLSocketImpl.connect(BaseSSLSocketImpl.java:173)
     at java.base/sun.net.NetworkClient.doConnect(NetworkClient.java:182)
     at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:474)
     at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:569)
     at java.base/sun.net.www.protocol.https.HttpsClient.<init>(HttpsClient.java:266)
     at java.base/sun.net.www.protocol.https.HttpsClient.New(HttpsClient.java:373)
     at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(AbstractDelegateHttpsURLConnection.java:203)
     at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1187)
     at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1081)
     at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:189)
     at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1592)
     at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1520)
     at java.base/java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:527)
     at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:334)
     at com.appsmith.server.helpers.GitUtils.isRepoPrivate(GitUtils.java:63)
     at com.appsmith.server.services.ce.GitServiceCEImpl.lambda$getApplicationCountWithPrivateRepo$208(GitServiceCEImpl.java:2288)
```

In the code, I found this snippet that makes an incorrect assumption that the SSH and HTTPS port are on the same domain:
```
    public static String convertSshUrlToBrowserSupportedUrl(String sshUrl) {
        if (StringUtils.isEmptyOrNull(sshUrl)) {
            throw new AppsmithException(AppsmithError.INVALID_PARAMETER, "ssh url");
        }
        return sshUrl
                .replaceFirst(".*git@", "https://")
                .replaceFirst("(\\.[a-z]*):", "$1/")
                .replaceFirst("\\.git$", "");
    }
```

Also:

1. `GitUtils.isRepoPrivate` uses blocking IO for this call, which seems highly undesirable in a WebFlux system?
2. It would be a good idea to set a connection timeout, because in firewalled environments, there's a good chance HTTPS is blocked off without refusing the connection, which might still cause a timeout...

### Public Sample App

_No response_

### Version

1.6.18

### PR DESCRIPTION
**Pull request by AnaghHegde**
## Description

> It would be a good idea to set a connection timeout, because in firewall environments, there's a good chance HTTPS is blocked off without refusing the connection, which might still cause a timeout. We are currently using blocking IO call for checking the repo status which is highly undesirable in webflux environment. 
This PR fixes the above mentioned issues. 

Fixes #12721 #12407 

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

- Locally
- JUnit tests

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/GitUtils.java [hunk 1]
```diff
@@ -57,12 +63,23 @@ public static String getRepoName(String remoteUrl) {
...
+                })
+                .exchange()
+                .flatMap(response -> {
+                    if (response.statusCode().is2xxSuccessful()) {
+                        return Mono.just(Boolean.FALSE);
+                    } else {
+                        return Mono.just(Boolean.TRUE);
+                    }
+                })
+                .onErrorResume(throwable -> Mono.just(Boolean.TRUE));
```
Comment (abhvsn) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/GitUtils.java:82]: Should we also log the error here?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/GitServiceCEImpl.java [hunk 2]
```diff
@@ -733,8 +728,10 @@ public Mono<Application> connectApplicationToGit(String defaultApplicationId, Gi
                     final String applicationId = application.getId();
                     final String orgId = application.getOrganizationId();
                     try {
-                        return fileUtils.checkIfDirectoryIsEmpty(repoPath)
-                                .flatMap(isEmpty -> {
+                        return fileUtils.checkIfDirectoryIsEmpty(repoPath).zipWith(GitUtils.isRepoPrivate(browserSupportedUrl))
```
Comment (abhvsn) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/GitServiceCEImpl.java:731]: As we are using the isRepoPrivate at couple of times we should cache the response and use at both the places.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/GitServiceCEImpl.java [hunk 3]
```diff
@@ -1905,7 +1897,10 @@ public Mono<ApplicationImportDTO> importApplicationFromGit(String organizationId
                     });
 
                     return defaultBranchMono
-                            .flatMap(defaultBranch -> {
+                            .zipWith(GitUtils.isRepoPrivate(GitUtils.convertSshUrlToBrowserSupportedUrl(gitConnectDTO.getRemoteUrl())))
```
Comment (abhvsn) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/GitServiceCEImpl.java:1900]: Same here we should cache the result

**Review by abhvsn — COMMENTED** [review]
Overall LGTM just a minor nits, also can we update the existing testcases for this

File: app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/GitUtils.java (hunk 1 above)
Comment (AnaghHegde) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/GitUtils.java:82]: I don't think this error information will be useful.

**Review by abhvsn — APPROVED** [review]
LGTM

**Pull request merged**